File: usacoGuide/Chapter2/Section2/runround.py
"""
ID: 06jajun
LANG: PYTHON3
TASK: runround
"""
best = float('inf')
def main():
    fin = open('runround.in', 'r')
    fout = open('runround.out', 'w')
    m = int(fin.readline().strip())
    fin.close()
    digs = [False] * 9
    numl = []

    def check():
        curr = 0
        s = set()
        n = len(numl)
        for i in range(n):
            curr = (curr + numl[curr]) % n
            s.add(curr)
        if len(s) == n:
            return True
        return False

    def recurse():
        if numl == []:
            for i in range(1, 10):
                numl.append(i)
                digs[i - 1] = True
                recurse()
                digs[i - 1] = False
                numl.pop()
            return
        numi = int(''.join([str(c) for c in numl]))
        global best
        if numi > best: return
        if numi > m and check():
            best = numi

        for i in range(1, 10):
            if digs[i - 1]:
                continue
            numl.append(i)
            digs[i - 1] = True
            recurse()
            digs[i - 1] = False
            numl.pop()
    recurse()
    fout.write(str(best)+'\n')
    fout.close()
    return


if __name__ == '__main__':
    main()

File: usacoGuide/Chapter2/Section2/runround.cpp
/*
ID: 06jajun
TASK: runround
LANG: C++17
*/

#include <iostream>
#include <algorithm>
#include <vector>
#include <fstream>
#include <unordered_map>
#include <string>

using namespace std;

#define MXN 256
#define INF 1000000000

int m;
ifstream fin ("runround.in");
ofstream fout ("runround.out");

bool check(int x) {
  string xs = to_string(x);
  bool there[9];
  int tot = 0;
  for (int i=0; i < 9; i++) there[i] = true;
  int curr = 0;
  int currc = xs[0] - 48;
  while (there[curr]) {
    curr = (curr + currc) % xs.length();
    currc = xs[curr] - 48;
    tot++;
  }
  if (curr == 0 && tot == xs.length()) {
    return true;
  }
  return false;
}

vector<int> roundnums(int x=0) {
  vector<int> ret;
  if (x > m) {
    if (check(x))
      ret.push_back(x);
    return ret;
  }
  x = x * 10;
  for (int i=1; i < 10; i++) {
    x++;
    for (auto r : roundnums(x)) ret.push_back(r);
  }
  return ret;
}

int main() {
  fin >> m;
  vector<int> ans = roundnums();
  sort(ans.begin(), ans.end());
  int bot = 0, top = ans.size() - 1;
  int end = top;
  int middle;
  while (bot <= top) {
    middle = (bot + top) / 2;
    if (ans[middle] < m) {
      bot = middle + 1;
    }
    else if (ans[middle] >= m) {
      end = middle;
      top = middle - 1;
    }
  }
  fout << ans[end] << endl;
}
